src/Main.purs

```diff
@@ -0,0 +1,529 @@
+compile bc@(BackendConfig backend) = do
+  JQuery.select "#column2" >>= \jq -> do
+    JQuery.empty jq
+    JQuery.create "<div>" >>= \div -> do
+      JQuery.addClass "loading" div
+      JQuery.appendText "Loading..." div
+      JQuery.append div jq
+
+  code <- fold <$> (JQuery.select "#code_textarea" >>= JQuery.getValueMaybe)
+
+  let displayPlainText s =
+        JQuery.select "#column2" >>= \jq -> do
+          JQuery.empty jq
+          JQuery.create "<pre>" >>= \pre -> do
+            JQuery.create "<code>" >>= \code_ -> do
+              JQuery.append code_ pre
+              JQuery.appendText s code_
+              JQuery.append pre jq
+
+  runContT (runExceptT (API.compile bc code)) \res_ ->
+    case res_ of
+      Left err -> displayPlainText err
+      Right res -> do
+        cleanUpMarkers
+
+        case res of
+          Right (CompileSuccess (SuccessResult { js })) -> do
+            showJs <- JQuery.select "#showjs" >>= \jq -> JQuery.is jq ":checked"
+            if showJs
+              then displayPlainText js
+              else backend.bundleAndExecute (JS js) bc
+          Right (CompileFailed (FailedResult { error })) ->
+            case error of
+              CompilerErrors errs -> do
+                JQuery.select "#column2" >>= JQuery.empty
+                forWithIndex_ errs \i (CompilerError{ message, position: ErrorPosition pos }) -> do
+                  JQuery.select "#column2" >>= \jq -> do
+                    h1 <- JQuery.create "<h1>"
+                    JQuery.addClass "error-banner" h1
+                    JQuery.appendText ("Error " <> show (i + 1) <> " of " <> show (Array.length errs)) h1
+
+                    pre <- JQuery.create "<pre>"
+                    code_ <- JQuery.create "<code>"
+                    JQuery.append code_ pre
+                    JQuery.appendText message code_
+
+                    JQuery.append h1 jq
+                    JQuery.append pre jq
+
+                  runEffFn4 addErrorMarker
+                    pos.startLine
+                    pos.startColumn
+                    pos.endLine
+                    pos.endColumn
+              OtherError err -> displayPlainText err
+          Left errs -> do
+            displayPlainText "Unable to parse the response from the server"
+            traverse_ (error <<< renderForeignError) errs
+
+-- | Set up a fresh iframe in the specified container, and use it
+-- | to execute the provided JavaScript code.
+foreign import setupIFrame
+  :: forall eff
+   . EffFn4 (dom :: DOM | eff)
+            JQuery.JQuery
+            String
+            String
+            (EffFn1 (dom :: DOM | eff) JQuery.JQuery Unit)
+            Unit
+
+-- | Execute the compiled code in a new iframe.
+execute :: forall eff. JS -> JS -> BackendConfig -> Eff (dom :: DOM | eff) Unit
+execute js bundle bc@(BackendConfig backend) = do
+  let html = joinWith "\n"
+        [ "<!DOCTYPE html>"
```

Comment:
This could be a bit more readable and easier to change with `"""` instead

Reply:
Done, thanks.